test/lifecycle_test.dart

```diff
@@ -175,6 +175,86 @@ main() {
 
         sharedTypeTests(testSnapshotType);
       });
+
+      test('triggers error lifecycle events when an error is thrown', () {
+        var mountNode = new DivElement();
+        var renderedInstance = react_dom.render(components2.SetStateTest({}), mountNode);
+        LifecycleTestHelper component = getDartComponent(renderedInstance);
+        LifecycleTestHelper.staticLifecycleCalls.clear();
+        component.setState({"shouldThrow": true});
+
+        // First render throws an error, caught by `getDerivedStateFromError`.
+        // `getDerivedStateFromError` sets state, starting the update cycle
+        // again. `componentDidCatch` also sets the state (storing error
+        // information), starting the update cycle again.
+        expect(
+            component.lifecycleCalls,
+            equals([
+              'shouldComponentUpdate',
+              'render',
+              'getDerivedStateFromError',
+              'shouldComponentUpdate',
+              'render',
+              'getSnapshotBeforeUpdate',
+              'componentDidUpdate',
+              'componentDidCatch',
+              'shouldComponentUpdate',
+              'render',
+              'getSnapshotBeforeUpdate',
+              'componentDidUpdate'
+            ]));
+        expect(component.state["shouldThrow"], isFalse,
+            reason: 'applies the state returned by `getDerivedStateFromError`');
+      });
+
+      test('can skip methods passed into _registerComponent2', () {
+        var mountNode = new DivElement();
+        var renderedInstance = react_dom.render(components2.SkipMethodsTest({}), mountNode);
+        LifecycleTestHelper component = getDartComponent(renderedInstance);
+        LifecycleTestHelper.staticLifecycleCalls.clear();
+        component.setState({"shouldThrow": true});
+
+        expect(
+            component.lifecycleCalls,
+            equals([
+              'shouldComponentUpdate',
+              'render',
+              'getDerivedStateFromError',
+              'shouldComponentUpdate',
+              'render',
+              'componentDidUpdate',
+              'componentDidCatch',
+              'shouldComponentUpdate',
+              'render',
+              'componentDidUpdate'
+            ]));
+      });
+
+      test('passes the correct error/info to lifecycle methods when an error is thrown', () {
+        int lengthOfKey = 15;
+        var mountNode = new DivElement();
+        var renderedInstance = react_dom.render(components2.SetStateTest({}), mountNode);
+        LifecycleTestHelper component = getDartComponent(renderedInstance);
+        Element renderedNode = react_dom.findDOMNode(renderedInstance);
+        LifecycleTestHelper.staticLifecycleCalls.clear();
+        component.setState({"shouldThrow": true});
+
+        expect(renderedNode.children[1].text.contains(getComponent2ErrorMessage()), isTrue);
```

Comment:
I would recommend the contains matcher; it gives you much more helpful error messages when this fails
```suggestion
        expect(renderedNode.children[1].text, contains(getComponent2ErrorMessage()));
```
Error message when using `isTrue`:
```
Expected: true
  Actual: <false>
```
Error message when using `contains`:
```
Expected: contains 'bar'
  Actual: 'foo'
```